core/templates/cowdata.h

```diff
@@ -200,43 +202,44 @@ void CowData<T>::_unref(void *p_data) {
 		return;
 	}
 
-	SafeNumeric<uint32_t> *refc = _get_refcount();
+	SafeNumeric<uint64_t> *refc = _get_refcount();
 
 	if (refc->decrement() > 0) {
 		return; // still in use
 	}
 	// clean up
 
 	if (!std::is_trivially_destructible<T>::value) {
-		uint32_t *count = _get_size();
+		uint64_t *count = _get_size();
 		T *data = (T *)(count + 1);
 
-		for (uint32_t i = 0; i < *count; ++i) {
+		for (uint64_t i = 0; i < *count; ++i) {
 			// call destructors
 			data[i].~T();
 		}
 	}
 
 	// free mem
-	Memory::free_static((uint8_t *)p_data, true);
+	Memory::free_static(((uint8_t *)p_data) - ALLOC_PAD, false);
 }
 
 template <class T>
-uint32_t CowData<T>::_copy_on_write() {
+uint64_t CowData<T>::_copy_on_write() {
 	if (!_ptr) {
 		return 0;
 	}
 
-	SafeNumeric<uint32_t> *refc = _get_refcount();
+	SafeNumeric<uint64_t> *refc = _get_refcount();
 
-	uint32_t rc = refc->get();
+	uint64_t rc = refc->get();
 	if (unlikely(rc > 1)) {
 		/* in use by more than me */
-		uint32_t current_size = *_get_size();
+		uint64_t current_size = *_get_size();
 
-		uint32_t *mem_new = (uint32_t *)Memory::alloc_static(_get_alloc_size(current_size), true);
+		uint64_t *mem_new = (uint64_t *)Memory::alloc_static(_get_alloc_size(current_size) + ALLOC_PAD, false);
```

Comment:
`Memory::alloc_static()` adds a 16-byte pre-padding, of which the last 8 bytes are availale for "users". When `CowData` was 32-bit, those 8 bytes were enough to store the two header `uint32_t`s. With this PR, `CowData` is adding 16 more bytes, which gives a total of 24 extra bytes. Assuming the ref count has to become 64-bit as well, we're wasting 8 bytes.

Possible options:
- a) Redefine `PAD_ALIGN` in `memory.h` to `24` and remove the newly added `ALLOC_PAD` in `CowData`.
- b) Redefine `PAD_ALIGN` in `memory.h` to `8` (assumming the extra bytes aren't used elsewhere) and let `CowData` deal with all the pre-padding (i.e., keep the current `ALLOC_PAD` of `16`).

Reply:
Only on debug builds, it doesn't add the padding otherwise, as the second parameter is false

Reply:
Note that the 8 bytes are used already, forget where but I've seen it, will look

Reply:
Yes I was going to mention this, it only adds 16 bytes on debug builds, but not on release, where it remains 16.

Reply:
I missed the argument has been changed to `false`. So, yeah, that's OK. Let's just find out (as a separate endeavor) if the extra pre-padding from `memory.h` is still needed somewhere.

Reply:
It is, in `memnew_arr` 🙂